### DISCUSSION TITLE
How to kill a thread that is running in Dispatcher.UIThread.Post ?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [AvaloniaUI/Avalonia#13427] [Q&A]
This is my code:
`RunBtn.Click += (sender, e) => Dispatcher.UIThread.Post(() => RunTheTask(), DispatcherPriority.Background);`

In the RunTheTask() function, I make a request to an API and wait for a response. However, there are situations where the process might hang due to network problems or other issues. 

How can I stop the frozen thread by clicking a 'stop' button?

**Comment by kyuranger** [comment]
It seems you can do it like this:
```
public MainWindow()
{
    InitializeComponent();
    RunBtn.Click += RunBtn_Click;            
}

private async void RunBtn_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
{
   //request to an API
}
```

**Reply by IronEaterBeast** [reply]
I don't understand, how to terminate the process when the API freezes.

**Reply by thevortexcloud** [reply]
This does not terminate the process. Rather it lets you run it async properly. Events in .NET are allowed to have `async void` handlers (it's really the only place you are meant to use them). This means if you await your process it will raise exceptions properly and you can then decide what to do when it fails.

However to answer your question, typically you would use some sort of cancellation token which allows for you to gracefully exit the process when something else happens.

https://learn.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads

But it's arguable if that is what you should be doing here.

**Comment by StefanKoell** [comment]
You can't really "kill" threads. If you are awaiting on a task, the way to handle it is to pass on a CancellationToken. If you start a thread manually and there's something running which blocks and doesn't support cancellation, you can only let it run and end by itself. Since it's running in background you can enable the UI again and let the user continue to work with your app. This heavily depends on the background code, of course.

**Reply by thevortexcloud** [reply]
If you are running your own threads directly (not using tasks) you can technically do it. It's just:

1. Not recommended
2. Does not work on modern versions of .NET

https://learn.microsoft.com/en-us/dotnet/standard/threading/destroying-threads

https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/5.0/thread-abort-obsolete

**Reply by IronEaterBeast** [reply]
Maybe I can use Process.Kill? I passed the CancellationToken, but the API didn't receive a response, which made it impossible to determine when to use the CancellationToken. As a result, the stop functionality didn't work.

**Reply by thevortexcloud** [reply]
Process.Kill only helps you if you are trying to kill an entire process (which I am guessing in this case would be your entire app). This is the equivalent of putting out a small kitchen fire with a tsunami.

**Comment by IronEaterBeast** [comment]
> You can't really "kill" threads. If you are awaiting on a task, the way to handle it is to pass on a CancellationToken. If you start a thread manually and there's something running which blocks and doesn't support cancellation, you can only let it run and end by itself. Since it's running in background you can enable the UI again and let the user continue to work with your app. This heavily depends on the background code, of course.

This is exactly the situation I encountered. I passed the CancellationToken, but the API didn't receive a response, which made it impossible to determine when to use the CancellationToken. As a result, the stop functionality didn't work.

**Reply by StefanKoell** [reply]
From my experience there are situations where CancellationTokens can't really cancel a long running task immediately. For example if a network call runs into a timeout you have to wait for the timeout first. The only way to deal with this is to leave it running in the background and not process the result.

**Reply by thevortexcloud** [reply]
You can put a timeout on your cancellation token so it will automatically cancel if nothing happens within a certain timeframe. 

https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/cancel-async-tasks-after-a-period-of-time

Typically though you pass the token as low down as you can get. If you are using `HttpClient` for example, all the async methods support taking a token.

**Reply by IronEaterBeast** [reply]
It seems I can set the timeout second with

`httpClient.Timeout = TimeSpan.FromSeconds(10);`

so it won't wait too long.

**Reply by thevortexcloud** [reply]
I suggest using a long timeout for HTTP requests. Some servers can be intermittently _very_ slow. Around ~3-5 minutes should be reasonable depending on what you are doing.
